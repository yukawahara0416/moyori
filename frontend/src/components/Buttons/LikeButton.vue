<template>
  <div class="mb-1">
    <v-btn
      fab
      small
      elevation="1"
      color="white"
      @click.stop="likeHandler(spot)"
      @mouseover="mouseover()"
      @mouseleave="mouseleave()"
    >
      <v-icon v-if="isLiking" color="error">mdi-heart</v-icon>
      <v-icon v-else :color="color">
        {{ icon }}
      </v-icon>
      <counter :spot="spot" :target="'likes'" />
    </v-btn>
  </div>
</template>

<script>
import { mapGetters, mapMutations, mapActions } from 'vuex'
import Counter from '@/components/Buttons/Counter.vue'
import { placeDetail, postSpot } from '@/plugins/maps.js'

export default {
  props: {
    spot: Object
  },

  components: {
    Counter
  },

  data() {
    return {
      icon: 'mdi-heart',
      color: '#757575'
    }
  },

  computed: {
    ...mapGetters(['form', 'map', 'headers', 'currentUser', 'isLoggingIn']),

    isLiking() {
      return this.yourLike.length > 0 ? true : false
    },

    yourLike() {
      return this.spot.hasYourVote('likes', this.currentUser.data.id)
    }
  },

  methods: {
    ...mapMutations(['setForm', 'dialogOn', 'changeSignTab']),
    ...mapMutations({ updateSpot: 'spot/updateSpot' }),
    ...mapActions([
      'vote',
      'unVote',
      'pushSnackbarSuccess',
      'pushSnackbarError'
    ]),

    likeHandler: async function(spot) {
      let newSpot = null
      const params = new FormData()

      try {
        if (!this.isLoggingIn) {
          this.changeSignTab('signin')
          this.dialogOn('dialogSign')
          throw new Error('ログインしてください')
        }

        // 未登録のスポットは登録します
        if (!spot.isPosted()) {
          newSpot = await this.getNewSpot(spot.data.place_id)
          params.append('like[spot_id]', newSpot.data.id)
        } else {
          params.append('like[spot_id]', spot.data.id)
        }

        this.voteHandler(newSpot || spot, params)

        this.pushSnackbarSuccess({ message: 'いいねしました' })
      } catch (error) {
        this.pushSnackbarError({
          message: error
        })
      }
    },

    getNewSpot: async function(place_id) {
      const updated = await placeDetail({ map: this.map, place_id })
      this.updateSpot({ place_id, updated })

      // formDataを用意してPOSTします
      this.setForm(this.spot)
      const newSpot = await postSpot(this.form, this.headers)
      this.updateSpot({ place_id, updated: newSpot })

      return newSpot
    },

    voteHandler: async function(spot, params) {
      const headers = this.headers
      const route = this.$route.name

      let isMyPage = false
      if (this.$route.params.id && this.currentUser.data.id) {
        isMyPage = this.$route.params.id == this.currentUser.data.id
      }

      // 「いいね」があれば「いいね」を取り消します
      if (this.isLiking) {
        const target = this.yourLike[0]
        await this.unVote({
          prop: 'likes',
          spot,
          target,
          headers,
          route,
          isMyPage
        })
        this.pushSnackbarSuccess({ message: 'いいねを取り消しました' })
        return
      }

      // 「いいね」します
      await this.vote({
        prop: 'likes',
        spot,
        params,
        headers,
        route,
        isMyPage
      })
    },

    mouseover() {
      this.icon = 'mdi-heart'
      this.color = 'error'
    },

    mouseleave() {
      this.icon = 'mdi-heart'
      this.color = '#757575'
    }
  }
}
</script>

<style scoped>
button::before {
  background-color: white;
}
</style>
