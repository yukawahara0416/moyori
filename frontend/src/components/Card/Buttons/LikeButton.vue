<template>
  <div>
    <v-btn icon @click.stop="likeHandler()">
      <v-icon v-if="isLiking" color="error">mdi-heart</v-icon>
      <v-icon v-else>mdi-heart-outline</v-icon>
      <counter :spot="spot" :genre="'likes'" />
    </v-btn>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'
import Counter from '@/components/Card/Buttons/Counter.vue'

export default {
  props: {
    spot: Object
  },

  components: {
    Counter
  },

  computed: {
    ...mapGetters(['currentUser', 'isLoggingIn', 'tab']),

    isPostedSpot() {
      return Object.prototype.hasOwnProperty.call(this.spot.data, 'id')
    },

    isLiking() {
      return this.likesByCurrentUser.length > 0 ? true : false
    },

    likesByCurrentUser() {
      if (this.spot.likes.length == 0) return []
      if (this.isLoggingIn == false) return []

      return this.spot.likes.filter(like => {
        return like.user_id == this.currentUser.data.id
      })
    }
  },

  methods: {
    ...mapActions({ saveSpot: 'map/saveSpot' }),
    ...mapActions(['like', 'unlike']),

    likeHandler: async function() {
      const spot = this.spot
      const activeTab = this.tab

      if (this.isLoggingIn == false) {
        this.dialogOn()
        this.$store.dispatch('pushSnackbar', {
          message: 'ログインしてください',
          color: 'error'
        })
        return
      }

      if (this.isPostedSpot == false) {
        const result = await this.saveSpot(spot)
        await this.like({ spot: result, active_tab: activeTab })
        return
      }

      if (this.isLiking == false) {
        await this.like({ spot: spot, active_tab: activeTab })
        return
      }

      await this.unlike({
        spot: spot,
        data: this.likesByCurrentUser[0],
        active_tab: activeTab
      })
    },

    dialogOn() {
      this.$store.dispatch('dialogOn', 'dialogSign')
    }
  }
}
</script>
