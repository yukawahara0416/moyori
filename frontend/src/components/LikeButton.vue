<template>
  <div>
    <v-btn icon @click="likeHandler()">
      <v-icon v-if="isLiked.length === 0">mdi-heart-outline</v-icon>
      <v-icon v-if="isLiked.length > 0" color="error">mdi-heart</v-icon>
      <p>{{ count }}</p>
    </v-btn>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'

export default {
  props: {
    spot: Object,
    id: Number
  },

  computed: {
    ...mapGetters(['spots', 'headers', 'currentUser']),

    isPosted() {
      if (this.spot.record.length !== 0) {
        return this.spot
      } else {
        return false
      }
    },

    isLiked() {
      const vm = this
      const likes = this.spots[this.id].likes
      if (vm.currentUser !== null && likes.length > 0) {
        const liked = likes.filter(function(like) {
          return like.user_id == vm.currentUser.data.id
        })
        return liked
      } else {
        return []
      }
    },

    count() {
      const likes = this.spots[this.id].likes
      if (likes.length > 0) {
        return likes.length
      } else {
        return 0
      }
    }
  },

  methods: {
    likeHandler: async function() {
      var spot = this.spot
      var id = this.id
      if (this.headers !== null) {
        if (this.isPosted) {
          if (this.isLiked.length === 0) {
            await this.like(spot, id)
          } else {
            await this.unlike(this.isLiked[0], id)
          }
        } else {
          spot = await this.$store.dispatch('postSpot', {
            spot: spot,
            id: id
          })
          await this.like(spot, id)
        }
      } else {
        this.$store.dispatch('pushSnackbar', {
          message: 'ログインしてください',
          color: 'error'
        })
      }
    },

    like(spot, id) {
      var params = { spot_id: spot.record.id }
      this.$store.dispatch('like', { params: params, id: id })
    },

    unlike(like, id) {
      var params = { id: like.id }
      this.$store.dispatch('unlike', { params: params, id: id })
    }
  }
}
</script>

<style></style>
