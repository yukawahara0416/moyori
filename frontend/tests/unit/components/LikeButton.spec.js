import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Vuetify from 'vuetify'
import LikeButton from '@/components/LikeButton.vue'
import spotStore from '@/store/modules/spot.js'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuetify)

let wrapper
let propsData
let store
let actions
let vuetify

beforeEach(() => {
  propsData = {
    spot: { record: { id: 1 } }
  }

  actions = {
    like: jest.fn(),
    unlike: jest.fn()
  }

  store = new Vuex.Store({
    modules: {
      likeStore: {
        actions
      },
      spotStore: {
        getters: spotStore.getters
      }
    }
  })

  vuetify = new Vuetify()

  wrapper = mount(LikeButton, {
    localVue,
    propsData,
    store,
    vuetify
  })
})

afterEach(() => {
  wrapper.destroy()
})

describe('props', () => {
  it('spot', () => {
    expect(wrapper.props().spot).toStrictEqual(propsData.spot)
    expect(wrapper.props().spot instanceof Object).toBe(true)
  })
})

describe('v-on', () => {
  it('likeHandler', () => {
    const event = jest.fn()
    wrapper.setMethods({ likeHandler: event })
    wrapper
      .findAll('.v-btn')
      .at(0)
      .trigger('click')
    expect(event).toHaveBeenCalledTimes(1)
  })
})

// describe('actions', () => {
//   it('like', () => {
//     wrapper.vm.like()
//     expect(actions.like).toHaveBeenCalled()
//   })

//   it('unlike', () => {
//     wrapper.vm.unlike()
//     expect(actions.unlike).toHaveBeenCalled()
//   })
// })

describe('template', () => {
  it('snapshot', () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })
})
