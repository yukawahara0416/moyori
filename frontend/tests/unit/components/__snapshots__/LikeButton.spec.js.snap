// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template snapshot 1`] = `
<div>
  <button
    class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
    type="button"
  >
    <span
      class="v-btn__content"
    >
      
    いいね
  
    </span>
  </button>
</div>
`;
